### ISSUE TITLE
SHELL is forgotten when ENV or ARG follows it (regression in 2.15.0)

### ISSUE THREAD

**Issue by rjaegers** [hadolint/hadolint#1252]
### Description

Since 2.15.0, hadolint reverts to assuming `/bin/sh` for `RUN` instructions if an `ENV` or `ARG` instruction appears between the `SHELL` instruction and the `RUN`. This produces false-positive POSIX-sh ShellCheck warnings (`SC3009`, `SC3014`, ...) for Dockerfiles that legitimately use bash.

`WORKDIR` and `COPY` in the same position do **not** trigger it, so it looks specific to the instructions that update the environment state.

### Reproduction

```dockerfile
FROM ubuntu:24.04
SHELL ["/bin/bash", "-Eeuo", "pipefail", "-c"]
ENV FOO=bar
RUN if [[ "$FOO" == "b" ]]; then echo hi; fi
```

```console
$ hadolint Dockerfile
Dockerfile:4 SC3014 warning: In POSIX sh, == in place of = is undefined.
```

Moving `SHELL` below the `ENV` makes the warning disappear. Replacing `ENV FOO=bar` with `ARG FOO=bar` reproduces it as well; replacing it with `WORKDIR /tmp` does not.

### Expected behaviour

`SHELL` applies to every subsequent `RUN` in the stage regardless of what other instructions come in between, so the snippet above should lint cleanly (as it did up to 2.14.0).

### Version bisect

| Version | Result |
| --- | --- |
| 2.12.0 | clean |
| 2.13.1 | clean |
| 2.14.0 | clean |
| 2.15.0 | `SC3014` |
| 2.15.1 | `SC3014` |

Platform: `hadolint-Linux-x86_64` release binaries, no config file.

Possibly introduced by #1180 ("Fix: propagating shell options across build stages"), which is in the 2.15.0 release notes -- though I have not confirmed that.

**Comment by m-ildefons** [comment]
I believe this is a duplicate of issue https://github.com/hadolint/hadolint/issues/1241
And that issue also is a duplicate of issue https://github.com/hadolint/hadolint/issues/1228 , which is fixed on `master` branch already.
